### PR TITLE
Speed up cold start

### PR DESCRIPTION
**Pull request by kaituo**
### Description
If historical data is enough, a single stream detector takes 1 interval for cold start to be triggered + 1 interval for the state document to be updated. Similar to single stream detectors, HCAD cold start needs 2 intervals and one more interval to make sure an entity appears more than once. So HCAD needs three intervals to complete cold starts. Long initialization is the single most complained problem of AD. This PR reduces both single stream and HCAD detectors' initialization time to 1 minute by
* delaying real time cache update by one minute when we receive ResourceNotFoundException in single stream detectors or when the init progress of HCAD real time cache is 0. Thus, we can finish the cold start and writing checkpoint one minute later and update the state document accordingly. This optimization saves one interval to wait for the state document update.
* disable the door keeper by default so that we won't have to wait an extra interval in HCAD.
* trigger cold start when starting a real time detector. This optimization saves one interval to wait for the cold start to be triggered.

Testing done:
* verified the cold start time is reduced to 1 minute.
* added tests for new code.

Signed-off-by: Kaituo Li <kaituo@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kaituo** [comment]
BWC tests failed:

 * What went wrong:
Execution failed for task ':adBwcCluster#mixedClusterTask'.
> `cluster{::adBwcCluster0}` failed to wait for cluster health yellow after 40 SECONDS
    IO error while waiting cluster
    503 Service Unavailable

Known issue: https://github.com/opensearch-project/OpenSearch/issues/5076

**Comment by amitgalitz** [comment]
>This PR reduces both single stream and HCAD detectors' initialization time to 1 minute by
Initial question, I see this PR has an improvement to reduce initialization time to just 1 minute not 1 interval + 1 minute. Why did we wait one interval before starting cold start? Does this mean if there is enough historical data, even if the interval is 10 minutes long, we can have a detector running in 1 minute?

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:148]: Previous PR had this extra 60 time wait when `response.getAnomalyScore() <= 0 || Double.isNaN(response.getAnomalyScore())) && response.getError() == null` was true. 
1. Is limiting this wait time to when init is 0 for hcad and when single stream throws `ResourceNotFoundException` narrowing the scope of this wait only happening when detector is first being created? 
2. Just to make sure we aren't going to get the discrepancy that sometimes happen between profile api and real time task api cause here we are internally checking if the realtime task already exists and we are above init of 0?

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:237]: Do single entity detectors also need 1 minute? if we restarted the detector is there any reason for us to believe initialization will be any quicker in the case we already have some results or when stopping detector we delete the model checkpoint along with any other relevant stored info?

**Comment by kaituo** [comment]
> > This PR reduces both single stream and HCAD detectors' initialization time to 1 minute by
> 
> Initial question, I see this PR has an improvement to reduce initialization time to just 1 minute not 1 interval + 1 minute. Why did we wait one interval before starting cold start? Does this mean if there is enough historical data, even if the interval is 10 minutes long, we can have a detector running in 1 minute?

Back then, I did manage to focus on improve cold start speed. This is long overdue.

Yes, even if the interval is 10 minutes long, we can have a detector running in 1 minute.

File: src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java:148]: yes, when the workflow reaches this line, detector index should exist. Otherwise, we will get index not found exception.

File: src/main/java/org/opensearch/ad/AnomalyDetectorPlugin.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/AnomalyDetectorPlugin.java:237]: The variable is only used in createComponents and can be created as a local variable instead of an instance variable.

File: src/main/java/org/opensearch/ad/settings/EnabledSetting.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/EnabledSetting.java:44]: created https://github.com/opensearch-project/anomaly-detection/issues/755

File: src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:699]: To catch the resource not found exception on the job scheduler and delay updating real time cache by 1 minute. Otherwise, we only get internal failure and has to process its error msgs.

File: src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java:134]: previously did we not load to cache? Cant find where this logic was previously if it was

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:123]: added comment:

Update real time task (one document per detector in state index). If the real-time task has no changes compared with local cache, the task won't update. Task only updates when the state changed, or any error happened, or AD job stopped. Task is mainly consumed by the front-end to track detector status. For single-stream detectors, we embed model total updates in AnomalyResultResponse and update state accordingly. For HCAD, we won't wait for model finishing updating before returning a response to the job scheduler since it might be long before all entities finish execution. So we don't embed model total updates in AnomalyResultResponse. Instead, we issue a profile request to poll each model node and get the maximum total updates among all models.

File: src/main/java/org/opensearch/ad/task/ADTaskManager.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADTaskManager.java:2827]: We don't need it since JVM will add .floatValue() automatically via unboxing. Prefer to keep it as it does not hurt and make the logic clear.

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:148]: 1. yes, you are right.
2. I haven't fixed the discrepancy issue. Tracked here: https://github.com/opensearch-project/anomaly-detection/issues/502 Plan to do it after the speed up PR is merged so the PR is not too big.

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:237]: yeah, 1 minute is to give enough leeway for training and checkpoint write, regardless of single stream or HCAD detectors.

For restart detector, as mentioned earlier, plan to do it after the speed up PR is merged so the PR is not too big. Currently, it is still an issue.

When stopping detector, we will delete everything except results. It will have to go through a new round of cold start.

File: src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java:134]: Previously, we don't load to cache when cold start finishes. We would have to wait for the next interval to make it happen.

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (amitgalitz) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:247]: sounds good

Comment (ohltyler) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:247]: For the restart scenario, since the checkpoint would have been deleted already, wouldn't it automatically go through this flow and have the 1 minute delay? In other words, what needs to be done differently in the restart scenario vs. the initial start scenario?

I may answer my own question as I keep reading

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:266]: should the realtime task error message be `e` instead of `errorMessage`? The latter will always be the one containing `CommonErrorMessages.NO_CHECKPOINT_ERR_MSG`. I'm not sure if it's more appropriate to keep that as the error message here, or if it's more useful to use the `e` error as returned by the polling failure. Maybe the `e` could be too unrelated to issues with the detector itself since it could be an issue with cluster, a particular node, etc.

File: src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java:127]: Can we add a debug log here that the detector failed to get added to cache after cold start?

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:265]: nit: typo in `execute`

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:247]: note it is restarting cluster not detector. So the checkpoint won't be deleted.

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:211]: can you add a comment here that it's not only indexing the result with the exception, but also updating the task state after 60s if the exception is related to cold start (index not found exceptions)?

File: src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java:123]: Just to clarify: this start listener is created & injected throughout the job handler so that whenever the job response is received, there's the extra step of trying to index results and update detector state with a 60s delay, correct?

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:266]: yeah, e could be too unrelated to issues with the detector itself since it could be an issue with cluster, a particular node, etc. errorMessage is the one due to running AnomalyResultAction and the real time cache mostly records that.

File: src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java:127]: The error is that we cannot find detector config from anomaly-detectors index, not failing to add it to cache after cold start. Is the error log enough?

File: src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java:344]: what is the purpose for having the listener on the stop detector job? currently the task state is already getting updated immediately after this action is finished. But with the new listener, wouldn't this trigger `indexAnomalyResult()` or `indexAnomalyResultException()`?

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:265]: fixed

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:211]: added

File: src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java:123]: yes

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:247]: it is a typo. Will change to cluster restart.

File: src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/EntityColdStartWorker.java:127]: Make sense. Changed to 

```
if (!detectorOptional.isPresent()) {
                        LOG.error(new ParameterizedMessage("fail to load trained model [{}] to cache due to the detector not being found.", modelState.getModelId()));
                        return;
                    }
```

File: src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java
Comment (amitgalitz) [reply src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java:133]: Thank you

**Review by amitgalitz — DISMISSED** [review]

**Review by ohltyler — APPROVED** [review]
LGTM - I think this will be a very visible improvement for users!

**Review by ohltyler — APPROVED** [review]

**Review by amitgalitz — APPROVED** [review]

**Pull request merged**
